pkg/services/ngalert/notifier/channels/util.go

```diff
@@ -37,10 +38,56 @@ const (
 
 var (
 	// Provides current time. Can be overwritten in tests.
-	timeNow              = time.Now
-	ErrImagesUnavailable = errors.New("alert screenshots are unavailable")
+	timeNow = time.Now
 )
 
+// firstImage returns the first image and the alert for which the image was
+// found. It returns nil if images are unavailable or if no images were found.
+func firstImage(ctx context.Context, images ImageStore, alerts ...*types.Alert) (*models.Image, *types.Alert, error) {
```

Comment:
I don't like the amount of duplication we see between this function and the `withStoredImage` family.

You can do this "first image" type call with a sentinel error in `withStoredImages`, which will abort iteration:
```go
_ = withStoredImages(ctx, sn.log, sn.images,
	func(index int, image *ngmodels.Image) error {
		if image != nil {
			req.Attachments[0].ImageURL = image.URL
			return fmt.Errorf("Found the first image") // Maybe use an error variable.
		}
		return nil
	}, alrts...)
```

So, you could write `firstImage` in terms of `withStoredImages` - keeping the function but reducing code duplication - or we could write that `withStoredImages` early exit pattern in notifiers where that makes sense.

Reply:
> You can do this "first image" type call with a sentinel error in withStoredImages, which will abort iteration:

That would work, I'm just not convinced it's the right thing to do here as just wanting the first image isn't an error?

Reply:
I think having the firstImage function is fine - then you can hide the sentinel error pattern in there. What I'm mainly concerned with is not re-writing all the iteration, store query, null-checking, etc logic.

It's not unheard of to use a sentinel error in Go code - `sql.NoRows` and `io.EOF`, for example.